Integrar ajustes para MDFe no repositório original da biblioteca DFe.NET

andrebts opened the pull request:
A biblioteca DFe.NET em seu módulo MDFe utiliza um conjunto de configurações estáticas que funcionam para aplicações com o contexto de apenas uma empresa. Entretanto, existe a necessidade em algumas aplicações de funcionar em um contexto multi-empresa.

A algum tempo atrás esses e outros ajustes foram feitos na [branch NET_Standard](https://github.com/andrebts/DFe.NET/tree/NET_Standard) de um fork pessoal. Entretanto, as mudanças que serão introduzidas pela nota técnica 2024.001 e futuras notas técnicas demandarão novas implementações. Então faz-se imperativo a unificação das modificações introduzidas na branch NET_Standard para a master do repositório original.

A alteração mais crucial foi a que ajustou a biblioteca para informar as configurações como instância e não estáticas (static) como eram. Como essas alterações nunca foram integradas ao repositório original, é esse comportamento foi reimplementado e a partir desse PR será integrado a branch master do repositório oficial.

andrebts committed:
Corrigido typo "Extencoes" para "Extensoes"

andrebts committed:
Adicionado singleton para MDFeConfiguracao;

andrebts committed:
Adicionado parametro opcional MDFeConfiguracao a ExtMDFe

andrebts committed:
Adicionado parâmetro MDFeConfiguracao a métodos das classes de extensão de MDFe;

Adicionado enum MDFeIndPag.

andrebts committed:
Adicionada descrição aos itens do enum MDFeModal;

Adicionado item RegimeEspecialNFF ao enum MDFeTipoEmissao;

Adicionado item PrestadorServicoDeTransporteCTeGlobalizado ao enum MDFeTipoEmitente;

Adicionado enum  MDFeTpCarga;

Adicionado item EncerramentoFisco ao enum MDFeTipoEvento;

Adicionado construtor em MDFeInfEvento para receber a versão do serviço;

Alterado nome de tpComp para MDFeTpComp e movido de MDFe.Classes.Informacoes para MDFe.Classes.Flags

andrebts committed:
Implementado INotifyPropertyChanged em MDFeConfiguracao;

Alterado nome de infMDFeSupl para MdfeInfMDFeSupl;

andrebts committed:
Renomeado Comp para MDFeComp;

Renomeado tpValePed para MDFeTpValePed;

andrebts committed:
Adicionado tratamento para obtenção do valor do campo dhIniViagem em MDFeIde a depender da versão do leiaute;

Renomeado infBanc para MDFeInfBanc e adicionado campo PIX;

Adicionado enum MDFeIndAltoDesemp;

Renomeado infPag para MDFeInfPag;

Renomeado infPrazo para MDFeInfPrazo;

andrebts committed:
Renomeado infEntregaParcial para MDFeInfEntregaParcial;

Renomeado infLocalCarrega para MDFeInfLocalCarrega;

Renomeado infLocalDescarrega para MDFeInfLocalDescarrega;

Renomeado infLotacao para MDFeInfLotacao;

Renomeado infEntregaParcial para InfEntregaParcial em MDFeInfCTe;

Renomeado xBalsa para XBalsa em MDFeInfEmbComb;

andrebts committed:
Renomeado prodPred para MDFeProdPred;

Renomeado infRespTec para MDFeInfRespTec;

andrebts committed:
Ajustado nome dos campos e adicionada documentação em MDFeInfRespTec.cs;

andrebts committed:
Alterada versão do leiaute específico para o Modal (VersaoModal) no construtor de MDFeInfModal da versão 1.00 para a versão 3.00;

Ajustada nomenclatura das propriedades de evPagtoOperMDFe para o estilo adotado nas outras classes;

Movida classe infViagens dentro de evPagtoOperMDFe.cs para a classe MDFeInfViagens em arquivo separado

andrebts committed:
Ajustado estilo das propriedades em MDFeValePed;

Ajustada propriedade UF de MDFeVeicReboque para permitir valores nulos uma vez que o campo é opcional.

andrebts committed:
Consolidado enum que representa "tpCarga";

Renomeada classe infContratante para MDFeInfContratante e adicionada descrição dos campos para manter o padrão de estilo das classes de MDFe;

Ajustado campo UF de MDFeVeicTracao para permitir valores nulos conforme Nota Tecnica 2021.001

andrebts committed:
Ajsutado nome da propriedade ProtMdFe para ProtMDFe da classe MDFeRetConsReciMDFe;

Adicionados campos de margem em ConfiguracaoDamdfe

andrebts committed:
Renomeado evPagtoOperMDFe para MDFeEvPagtoOperMDFe afim de manter compatibilidade do estilo utilizado no pacote MDFe;

Adicionado parâmetro MDFeConfiguracao ao método MDFeConsultaNaoEncerrados em ServicoMDFeConsultaNaoEncerrados;

Adicionado parametro e tratamento para configuração em métodos de ClassesFactory

andrebts committed:
Adicionado parâmetro MDFeConfiguracao em  ServicoMDFeConsultaProtocolo.MDFeConsultaProtocolo e IServicoController.Executar;

Adicionado parâmetro MDFeConfiguracao  em FactoryEvento.CriaEvento;

Adicionado parâmetro MDFeConfiguracao  em diversos métodos de WsdlFactory;

andrebts committed:
Adicionado parâmetro opcional MDFeConfiguracao em métodos das classes EventoCancelar, EventoEncerramento, EventoInclusaoCondutor, EventoInclusaoDFe, EventoPagamentoOperacao, ServicoMDFeEvento, ServicoMDFeRecepcao, ServicoMDFeRetRecepcao e ServicoMDFeStatusServico.

andrebts committed:
Movida classe MDFeVersaoWebService para seu próprio arquivo

andrebts committed:
Adicionado tratamento para CPF ao método ObterChave de ChaveFiscal;

andrebts committed:
Adicionado pacote de testes para MDFe;

Adicionados testes unitários para MDFe;

andrebts committed:
Removido pacote MDFe.Tests

andrebts committed:
Excluidos arquivos do pacote MDFe.Tests

andrebts committed:
Ajustados nomes de propriedades para estarem conforme estilo utilizado no restante do pacote

andrebts committed:
Adicionado tratamento para casos em que a data e hora de início da viagem é nulo;

Adicionada verificação para serialização do campo IE (Inscrição Estadual);

Adicionado tratamento para geração de QRCode ao assinar MDFe

andrebts committed:
Ajsutado nome da propriedade InfANTT de MDFeRodo para estar em conformidade com o padrão utilizado nos demais arquivos do projeto MDFe

andrebts committed:
Adicionados métodos ExportarPdf e ObterHTML a DamdfeFrMDFe;

Ajustado arquivo de relatório do Fast Report para refletir as mudanças no nome das propriedades das entidades;

Ajustado arquivo de relatório do Fast Report para evitar erro CS0117 pelo uso de Environment.NewLine diretamente no relatório;

andrebts committed:
Ajustado método SalvarXmlEmDisco para avaliar a configuração a ser utilizada;

Adicionado método ShouldSerialize a IndAltoDesemp para serializar apenas quando seu valor é AltoDesempenho;

Configurados arquivos de esquema evPagtoOperMDFe_v3.00.xsd e retMDFe_v3.00.xsd para serem copiados quando houver uma nova versão.

AgnaldoSilva0 commented on MDFe.Classes/Extensoes/ExtMDFe.cs:
```diff
@@ -166,12 +178,14 @@ public static string XmlString(this MDFEletronico mdfe)
+        public static void SalvarXmlEmDisco(this MDFEletronico mdfe, string nomeArquivo = null, MDFeConfiguracao cfgMdfe = null)
         {
-            if (MDFeConfiguracao.NaoSalvarXml()) return;
+            if (MDFeConfiguracao.Instancia.NaoSalvarXml()) return;
```
Em alguns outros métodos, a instância é verificada usando a configuração recebida por parâmetro. Caso esta seja nula, é então utilizada a instância geral estática. Entretanto, neste método específico, a instância geral está sendo utilizada diretamente, independentemente de ter sido fornecida uma por parâmetro. Assim, é necessário ajustá-lo para dar prioridade à configuração recebida.

eldermatheus commented on MDFe.Classes/Informacoes/MDFeInfRespTec.cs:
```diff
@@ -0,0 +1,90 @@
+        [XmlIgnore]
+        private int? idCSRT { get; set; }
+
+        public bool idCSRTSpecified
```
Acredito que essa variável deveria começar com letra maiúscula

eldermatheus commented on MDFe.Damdfe.Fast/DamdfeFrMDFe.cs:
```diff
@@ -164,5 +165,81 @@ public void ExportarPdf(Stream outputStream, FastReport.Export.ExportBase export
+            stream.Position = 0;
+
+            return stream;
+
```
Remover essa linha sem necessidade

AgnaldoSilva0 commented on MDFe.Damdfe.OpenFast/DamdfeFrMDFe.cs:
```diff
@@ -75,13 +76,19 @@ public void RegisterData(MDFeProcMDFe proc)
         }
 
-        public void Configurar(ConfiguracaoDamdfe config)
+       public void Configurar(ConfiguracaoDamdfe config)
```
Dê um espaço para o lado nessa linha, note que está desalinhada.

AgnaldoSilva0 commented on MDFe.Servicos/ConsultaNaoEncerradosMDFe/ServicoMDFeConsultaNaoEncerrados.cs:
```diff
@@ -31,21 +31,22 @@
+            var webService = WsdlFactory.CriaWsdlMDFeConsNaoEnc(cfgMdfe);
             var retornoXml = webService.mdfeConsNaoEnc(consMDFeNaoEnc.CriaRequestWs());
 
             var retorno = MDFeRetConsMDFeNao.LoadXmlString(retornoXml.OuterXml, consMDFeNaoEnc);
```
Na chamada de método abaixo `retorno.SalvarXmlEmDisco(cnpjCpf);`, ele também recebe a configuração como parâmetro. Você acha ideal usar a configuração nesse método, como foi feito nos outros acima?

Ficaria assim: `retorno.SalvarXmlEmDisco(cnpjCpf, cfgMdfe);` (Linha 53)

AgnaldoSilva0 commented on MDFe.Servicos/ConsultaProtocoloMDFe/ServicoMDFeConsultaProtocolo.cs:
```diff
@@ -31,21 +31,22 @@
+            var webService = WsdlFactory.CriaWsdlMDFeConsulta(cfgMdfe);
             var retornoXml = webService.mdfeConsultaMDF(consSitMdfe.CriaRequestWs());
 
             var retorno = MDFeRetConsSitMDFe.LoadXml(retornoXml.OuterXml, consSitMdfe);
```
Aqui também é o mesmo caso do comentário anterior, a chamada na linha 53 também pode receber a configuração como parâmetro.

Ficaria assim: `retorno.SalvarXmlEmDisco(chave, cfgMdfe);` (Linha 53)

AgnaldoSilva0 commented on MDFe.Servicos/EventosMDFe/EventoPagamentoOperacao.cs:
```diff
@@ -1,24 +1,62 @@
+
+using System.Collections.Generic;
 using MDFe.Classes.Informacoes;
 using MDFe.Classes.Informacoes.Evento.CorpoEvento;
```
`using` sem uso.

AgnaldoSilva0 commented on MDFe.Utils/Configuracoes/MDFeConfiguracao.cs:
```diff
@@ -38,41 +38,85 @@
 using DFe.Utils.Assinatura;
 using System;
 using VersaoServico = MDFe.Utils.Flags.VersaoServico;
```
`using's` sem uso:

```
using DFe.Classes.Entidades;
using DFe.Classes.Flags;
using VersaoServico = MDFe.Utils.Flags.VersaoServico;
```

AgnaldoSilva0 commented on MDFe.Utils/Configuracoes/MDFeConfiguracao.cs:
```diff
@@ -84,16 +128,59 @@ public static X509Certificate2 X509Certificate2
+            if (handler != null) handler(this, new PropertyChangedEventArgs(propertyName));
+        }
+
         public void Dispose()
```
Não faz parte do código modificado, mas como essa classe sofreu alterações significativas e seguindo o princípio do escoteiro, poderia verificar o seguinte código:

```
public void Dispose()
{
    if (!ConfiguracaoCertificado.ManterDadosEmCache && _certificado != null)
    {
        _certificado.Reset();
        _certificado = null;
    }
}

~MDFeConfiguracao()
{
    if (!ConfiguracaoCertificado.ManterDadosEmCache && _certificado != null)
    {
        _certificado.Reset();
        _certificado = null;
    }
}
```

Note que o destrutor faz a mesma coisa do Dispose, assim, visando remover a duplicação de código, poderia analisar a possibilidade de encapsular a lógica de liberação de recursos em um método, onde, tanto o dispose, como o destrutor chame este método.

Sugestão (Pode ajustar conforme achar necessário): 
```
public void Dispose()
{
    LiberarRecursos();
}

~MDFeConfiguracao()
{
    LiberarRecursos();
}

private void LiberarRecursos()
{
    if (NaoDeveManterDadosDoCertificadoEmCache())
    {
        _certificado.Reset();
        _certificado = null;
    }
}

private bool NaoDeveManterDadosDoCertificadoEmCache()
    => !ConfiguracaoCertificado.ManterDadosEmCache && _certificado != null;
```

AgnaldoSilva0 commented on DFe.Utils/ChaveFiscal.cs:
```diff
@@ -59,6 +59,11 @@ public static DadosChaveFiscal ObterChave(Estado ufEmitente, DateTimeOffset data
+                cnpjEmitente = cnpjEmitente.PadLeft(14, '0');
+            }
+
             chave.Append(((int)ufEmitente).ToString("D2"))
```
Aproveite que fez alterações nesse método, e alinhe os `Append`.

```
chave.Append(((int)ufEmitente).ToString("D2"))
     .Append(dataEmissao.ToString("yyMM"))
     .Append(cnpjEmitente)
     .Append(((int)modelo).ToString("D2"))
     .Append(serie.ToString("D3"))
     .Append(numero.ToString("D9"))
     .Append(tipoEmissao.ToString())
     .Append(cNf.ToString("D8"));
```

AgnaldoSilva0 commented on MDFe.Classes/Extensoes/ExtMDFe.cs:
```diff
@@ -72,63 +74,65 @@ public static MDFEletronico Valida(this MDFEletronico mdfe)
 
```
Remover linha desnecessária. (73 ou 74)

AgnaldoSilva0 commented on MDFe.Classes/Extensoes/ExtMDFeRetConsSitMDFe.cs:
```diff
@@ -36,15 +36,17 @@
+
+            var caminhoXml = config.CaminhoSalvarXml;
 
             var arquivoSalvar = Path.Combine(caminhoXml,  chave + "-sit.xml");
```
Remover espaço a mais entre a virgula e o parametro `chave`.

`var arquivoSalvar = Path.Combine(caminhoXml, chave + "-sit.xml");`

AgnaldoSilva0 commented on MDFe.Classes/Flags/MDFeTpCarga.cs:
```diff
@@ -1,9 +1,12 @@
+    public enum MDFeTpCarga
     {
+        [XmlEnum("0")]
+        Indefinido = 0,
```
A tag `tpCarga` é obrigatória.
A inserção do tipo Indefinido 0 é por algum motivo especifico? Já que não se encontra na lista de possíveis valores válidos para a tag.

andrebts replied:
Não havia prestado atenção nele, o removi pois não há documentação que pude encontrar para embasar sua presença

AgnaldoSilva0 commented on MDFe.Classes/Flags/MDFeTpComp.cs:
Aproveite que fez alteração nesse enumerador, e corrija o nome: 
`DespesasBancariasEmiosDePagamentoOutras`
Para: 
`DespesasBancariasEMeiosDePagamentoOutras`

AgnaldoSilva0 commented on MDFe.Classes/Informacoes/MDFeDisp.cs:
```diff
@@ -42,33 +43,43 @@ public class MDFeDisp
+        /// <summary>
+        /// 4 - Valor do Vale-Pedagio 
+        /// </summary>
         public decimal vValePed
```
Para seguir o padrão de nomenclaturas, poderia ajustar o `vValePed ` para `VValePed`.
Se seguir a sugestão, acredito que deve indicar ao serializador qual tag se refere: 
`[XmlElement(ElementName = "vValePed")]`

AgnaldoSilva0 commented on MDFe.Classes/Informacoes/MDFeIde.cs:
```diff
@@ -226,11 +235,41 @@ public string ProxyUFFim
         /// </summary>
         [XmlElement(ElementName = "dhIniViagem")]
-        public string ProxyDhIniViagem { get; set; }
+        public string ProxyDhIniViagem {
```
Mover a abertura de chave do método para a linha de baixo.

AgnaldoSilva0 commented on MDFe.Classes/Informacoes/MDFeIde.cs:
```diff
@@ -226,11 +235,41 @@ public string ProxyUFFim
+                    default:
+                        throw new InvalidOperationException("Versão Inválida para MDF-e");
+                }
+
```
Remover linha desnecessária. (Linha 252)

AgnaldoSilva0 commented on MDFe.Classes/Informacoes/MDFeInfLocalCarrega.cs:
```diff
@@ -0,0 +1,82 @@
+            get
+            {
+                if (_latitude == null) return null;
+                    return _latitude.ToString();
```
Remover o recuo do `return _latitude.ToString();`. Da maneira como está, dá a impressão que está dentro do if, o que não é verdade.

Sugestão: 
```
if (_latitude == null) return null;

return _latitude.ToString();
```

AgnaldoSilva0 commented on MDFe.Classes/Informacoes/MDFeInfLocalCarrega.cs:
```diff
@@ -0,0 +1,82 @@
+        /// <summary>
+        /// 1 - Longitude do ponto geográfico onde foi carregado o MDF-e.
+        /// </summary>
+        [XmlElement("Longitude")]
```
A tag `longitude ` no xml é com letras minúsculas.

Ajustar para:
`[XmlElement("longitude")]`

AgnaldoSilva0 commented on MDFe.Classes/Informacoes/MDFeInfLocalDescarrega.cs:
```diff
@@ -0,0 +1,83 @@
+        /// 1 - Longitude do ponto geográfico onde foi
+        /// carregado o MDF-e.
+        /// </summary>
+        [XmlElement("Longitude")]
```
Mesmo caso anterior.

Ajustar para:
`[XmlElement("longitude")]`

AgnaldoSilva0 commented on MDFe.Classes/Informacoes/MDFeInfLocalDescarrega.cs:
```diff
@@ -0,0 +1,83 @@
+
+        /// <summary>
+        /// 1 - Longitude do ponto geográfico onde foi
+        /// carregado o MDF-e.
```
O comentário tem um pequeno erro, essa tag se refere ao local de descarregamento, mas no comentário se refere ao carregamento.

Ajustar para:
```
/// 1 - Longitude do ponto geográfico onde foi
/// descarregado o MDF-e.
```

AgnaldoSilva0 commented on MDFe.Classes/Informacoes/MDFeInfLocalDescarrega.cs:
```diff
@@ -0,0 +1,83 @@
+
+        /// <summary>
+        /// 1- Latitude do ponto geográfico onde foi
+        /// carregado o MDF-e.
```
O comentário tem um pequeno erro, essa tag se refere ao local de descarregamento, mas no comentário se refere ao carregamento.

Ajustar para:
```
/// 1- Latitude do ponto geográfico onde foi
/// descarregado o MDF-e.
```

AgnaldoSilva0 commented on MDFe.Classes/Informacoes/MDFeInfLocalDescarrega.cs:
```diff
@@ -0,0 +1,83 @@
+    public class MDFeInfLocalDescarrega
+    {
+        /// <summary>
+        /// 1 - Cep onde foi carregado o MDF-e.
```
O comentário tem um pequeno erro, essa tag se refere ao local de descarregamento, mas no comentário se refere ao carregamento.

Ajustar para:
`/// 1 - Cep onde foi descarregado o MDF-e.`

AgnaldoSilva0 commented on MDFe.Classes/Informacoes/MDFeInfMDFe.cs:
```diff
@@ -47,6 +47,7 @@ public MDFeInfMDFe()
             Tot = new MDFeTot();
+            ProdPred = new MDFeProdPred();
             Versao = VersaoServico.Versao100;
         }
```
Adicionar um espaço entre o final da chave do construtor e `/// <summary>`.
Mais especificamente entre as linhas 52 e 53.

AgnaldoSilva0 commented on MDFe.Classes/Informacoes/MDFeInfPag.cs:
```diff
@@ -0,0 +1,122 @@
+        public string CNPJ { get; set; }
+
+        /// <summary>
+        /// 3 - Componentes do Pagamento do Frete.
```
Comentário não condiz com a tag, esse comentário é referente a tag Comp.

Ajustar para:
```
/// 3 - Identificador do responsável pelo pgto 
/// em caso de ser estrangeiro
```

AgnaldoSilva0 commented on MDFe.Classes/Informacoes/MDFeInfPag.cs:
```diff
@@ -0,0 +1,122 @@
+        public MDFeInfBanc InfBanc { get; set; }
+
+        /// <summary>
+        /// 3 - Informações Bancárias.
```
Comentário não condiz com a tag, essa descrição é referente a tag InfBanc.

Ajustar para:
```
/// 3 - Indicador de operação de transporte de 
/// alto desempenho 
```

AgnaldoSilva0 commented on MDFe.Classes/Informacoes/MDFeInfPercurso.cs:
```diff
@@ -45,14 +46,12 @@ public class MDFeInfPercurso
+        /// Proxy para sigla das Unidades da Federação do percurso do veículo
         /// </summary>
         [XmlElement(ElementName = "UFPer")]
         public string ProxyUFPer {
```
Mover a abertura de chave do ProxyUFPer para a linha de baixo.

```
public string ProxyUFPer
{
    get { return UFPer.GetSiglaUfString(); }
    set { UFPer = UFPer.SiglaParaEstado(value); }
}
```

AgnaldoSilva0 commented on MDFe.Classes/Informacoes/MDFeProdPred.cs:
```diff
@@ -0,0 +1,80 @@
+/*  Esta biblioteca é distribuída na expectativa de que seja útil, porém, SEM   */
+/* NENHUMA GARANTIA; nem mesmo a garantia implícita de COMERCIABILIDADE OU      */
+/* ADEQUAÇÃO A UMA FINALIDADE ESPECÍFICA. Consulte a Licença Pública Geral Menor*/
+/* ADEQUAÇÃO A UMA FINALIDADE ESPECÍFICA. Consulte a Licença Pública Geral Menor*/
```
Remover linha repetida.

AgnaldoSilva0 commented on MDFe.Classes/Informacoes/MDFeVeicTracao.cs:
```diff
@@ -116,9 +117,13 @@ public string ProxyUF
+            set
+            {
+                if (!string.IsNullOrEmpty(value))
+                    UF = new Estado().SiglaParaEstado(value);
```
Na classe `MDFeVeicReboque` a propriedade `ProxyUF` existe o tratamento `else` setando null para UF em caso do valor informado ser vazio, mas aqui esse comportamento não foi usado. Tem algum motivo especifico?

andrebts replied:
Adicionei o tratamento também para MDFeVeicTracao.

andrebts committed:
Ajustado nome da propriedade IdCSRT da classe MDFeInfRespTec para atender aos padrões de estilo utilizados;

Adicionada passagem de parametro do tipo MDFeConfiguracao para chamadas ao método SalvarXmlEmDisco;

andrebts committed:
Refatorado conteúdo do método Dispose e do destrutor de MDFeConfiguracao para reduzir a repetição de código;

andrebts committed:
Removido valor "Indefinido" do Enum MDFeTpCarga;

Ajustado nome da propriedade VValePed da classe MDFeDisp para estar em conformidade com o padrão de estilo utilizado nos projetos;

andrebts committed:
Ajuste no nome do elemento XML da propriedade longitude das classes MDFeInfLocalCarrega e MDFeInfLocalDescarrega;

Ajustado summary de propriedades das classes MDFeInfLocalDescarrega e MDFeInfPag;

Adicionado tratamento para set da propriedade ProxyUF quando o valor informado for vazio ou nulo.
